backend/src/tests/demo_db/conftest.py

```diff
@@ -0,0 +1,159 @@
+import asyncio
+import time
+from contextlib import asynccontextmanager
+from typing import AsyncIterator
+
+import httpx
+import mirakuru
+import pytest
+from app.main import app
+from app.models import BaseDAO, UserDAO
+from sqlalchemy import create_engine
+from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
+from sqlalchemy.orm import sessionmaker
+
+# shell command to launch ephemeral *in-memory* cockroach db
+COCKROACH_CMD = """cockroach demo --sql-port 26259 --no-example-database -e 'ALTER USER demo with password "noteable";select pg_sleep(1000)'"""
+
+# these should match with the COCKROACH_CMD values
+TEST_DB_USER = "demo"
+TEST_DB_PASS = "noteable"
+TEST_DB_HOST = "localhost"
+TEST_DB_PORT = "26259"
+TEST_DB_NAME = "defaultdb"
+
+# TOC for this file:
+#   1. Override pytest-asyncio event_loop fixture to be session scoped
+#   2. Set up an async engine and db_session that will do SAVEPOINT restorations per test
+#   3. Fixture for creating a seed test user
+#   4. Fixture for creating an httpx.AsyncClient and client that is authorized as test user
+#   5. Fixture to start an in-memory cockroach db and create the tables
+#
+# Of those, only #5 is run automatically.
+# The rest are explicitly called as fixture arguments in tests
+
+# 1. override pytest-asyncio event_loop
+# If this isn't done, then tests will show errors like "another operation is in progress"
+# or "attached to a different loop"
+@pytest.fixture(scope="session")
+def event_loop():
+    return asyncio.get_event_loop()
+
+
+# 2. Set up async engine and db_session
+# The engine handles the SAVEPOINT and ROLLBACK
+# Sessions inside a .begin_nested() block that try to .commit()
+# produce RELEASE SAVEPOINT rather than COMMIT statements
+#
+# Note that patching db_session into route files (e.g. patch('app.auth.db_session', db_session) )
+# will be done in other files, typically with an autouse=True fixture.
+# Since db_session uses the engine prefix, you'll see BEGIN / SAVEPOINT / ROLLBACK
+# messages (if you display stdout) even when there is no database interaction in the actual test
+
+ASYNC_TEST_DB_URL = f"cockroachdb+asyncpg://{TEST_DB_USER}:{TEST_DB_PASS}@{TEST_DB_HOST}:{TEST_DB_PORT}/{TEST_DB_NAME}"
+async_engine = create_async_engine(ASYNC_TEST_DB_URL)
+
+
+@pytest.fixture
+async def engine():
+    async with async_engine.connect() as conn:
+        await conn.begin()
+        await conn.begin_nested()
+        yield conn
+        await conn.rollback()
+
+
+@pytest.fixture
+async def db_session(engine):
```

Comment:
Should annotate db_session()'s return type here, and probably through a declared TypeVar declared at the best outermost context so that would be usable throughout the codebase, both tests and main code. It looks like it is something more like Callable[AsyncIterator[AsyncSession]]. "AsyncSessionMaker"? English was not prepared for the noun explosion OOP has brought down upon us.

Looks like most places depending on db_session fixture and whatnot do not have it type annotated properly.

Welcome to Python, where the type annotations DO matter but they don't guarantee a thing.